### ISSUE TITLE
LLM

### ISSUE THREAD

**Issue by zhangjun** [zhangjun/zhangjun.github.io#32]
# repo链接
https://github.com/THUDM/ChatGLM-6B
https://github.com/mymusise/ChatGLM-Tuning
https://github.com/LianjiaTech/BELLE

## LLM量化
https://zhuanlan.zhihu.com/p/616969812
- [SmoothQuant](https://github.com/mit-han-lab/smoothquant)
- Outlier Suppression
   - [Outlier Suppression](https://arxiv.org/abs/2209.13325)
   - [Outlier Suppression+](https://arxiv.org/abs/2304.09145)
- [AWQ](https://arxiv.org/abs/2306.00978)
 基于激活与参数大小的缩放保护
- [LLM.int8](https://arxiv.org/abs/2208.07339)
   - LLM.int8(): 8-bit Matrix Multiplication for Transformers at Scale
   - https://timdettmers.com/2022/08/17/llm-int8-and-emergent-features/
   - https://github.com/timdettmers/bitsandbytes
- [GPTQ](https://arxiv.org/pdf/2210.17323.pdf)
   - [GPTQ: 模型量化，穷鬼救星](https://zhuanlan.zhihu.com/p/616969812)
      不使用贪心算法，对W进行优化时，固定位置挑选Q
      W不同列间权重更新互相独立，使用批处理更新（Lazy Batch更新，group_size参数控制）
      数值稳定性优化
   - [4-bit LLM Quantization with GPTQ](https://mlabonne.github.io/blog/posts/4_bit_Quantization_with_GPTQ.html)
- [ZeroQuant](https://arxiv.org/abs/2206.01861)
   - https://github.com/microsoft/DeepSpeed/pull/2217
   - [ZeroQuant-V2: Exploring Post-training Quantization in LLMs from Comprehensive Study to Low Rank Compensation](https://arxiv.org/abs/2303.08302)
- [LUT-GEMM](https://arxiv.org/pdf/2206.09557.pdf)
   - [LUT-GEMM: 基于lut的量化矩阵乘法在大规模生成语言模型中的高效推理](https://blog.csdn.net/weixin_42764932/article/details/131230429?spm=1001.2014.3001.5501)
- [SparseGPT](https://arxiv.org/pdf/2301.00774.pdf)
- weight only
   - [Who Says Elephants Can’t Run: Bringing Large Scale MoE Models into Cloud Scale Production](https://arxiv.org/pdf/2211.10017.pdf) 
   - [TVM weight-only](https://github.com/apache/tvm/pull/15111)
   - [cutlass_fpA_intB_gemm](https://github.com/tlc-pack/cutlass_fpA_intB_gemm/pull/1/files)

## fine-tune
- [PEFT](https://arxiv.org/abs/2106.09685)
- [QLora](https://arxiv.org/abs/2305.14314)

# LLM papers
[Awesome-LLM-System-Papers](https://github.com/AmadeusChan/Awesome-LLM-System-Papers)
[SpecInfer: Accelerating Generative LLM Serving with Speculative Inference and Token Tree Verification](https://www.cs.cmu.edu/~zhihaoj2/papers/specinfer.pdf)  
https://flexflow.ai/specInfer/
https://www.gabrieleoliaro.com/publication/expertflow/expertflow.pdf

**Comment by zhangjun** [comment]
# LLM Inference Optimization
https://lilianweng.github.io/posts/2023-01-10-inference-optimization/
https://www.nebuly.com/blog/full-stack-optimization-of-transformer-inference-a-survey-part-1
[Full Stack Optimization of Transformer Inference: a Survey](https://arxiv.org/pdf/2302.14017.pdf)
## continous batching
[continuous batching llm inference](https://www.anyscale.com/blog/continuous-batching-llm-inference)

**Comment by zhangjun** [comment]
# TensorRT & Triton
[Accelerated Inference for Large Transformer Models Using NVIDIA Triton Inference Server](https://developer.nvidia.com/blog/accelerated-inference-for-large-transformer-models-using-nvidia-fastertransformer-and-nvidia-triton-inference-server)

**Comment by zhangjun** [comment]
# 大模型分布式推理
## 推理方案
||<=百亿|百亿～千亿|>千亿|
|-|-|-|-|
|资源|单机多卡|单机多卡|多机多卡|
|并行策略|张量并行|张量并行|张量+流水线并行|
|通讯|单机all reduce|单机all reduce|单机all reduce+多机send/recv|
## 模型切分策略

**Comment by zhangjun** [comment]
## LLM应用
### LangChain
<img src="https://github.com/zhangjun/zhangjun.github.io/assets/1312389/19a1863d-4ae3-460c-902b-1a2b03b768d6" width="600">

**Comment by zhangjun** [comment]
# Flash Attention
## [Flash Attention](https://openreview.net/pdf?id=H4DqfPSibmx)
- parallelism
parallelize over batch_size and num of heads
flash attention2 - long sequences(small batch size or num of heads), parallelize over sequence length dimension
- better work partition
  reduce the amount of synchronization and communication between different warps
  FlashAttention splits K and V across 4 warps while keeping Q accessible by all warps.
  FlashAttention-2 splits Q across 4 warps while keeping K and V accessible by all warps.
  ![image](https://github.com/zhangjun/zhangjun.github.io/assets/1312389/703cdb9d-927b-4316-85ad-58d380b9478d)
- supported head dimensions up to 256 and MQA
   GQA、MQA
## Flash Attention2 优化
<img width="657" alt="image" src="https://github.com/zhangjun/zhangjun.github.io/assets/1312389/2cadba46-f1e4-4b4b-a2bc-aaa0c0061b4d">

Flash Attention2优化点详解
- Fused kernel与矩阵分块
- Causal Masking
- Non-Matmul 计算优化
- 流水编排与异步加载和Double Buffer
- Layout Swizzle

**Comment by zhangjun** [comment]
# TensorRT-LLM
streaming of tokens, in-flight batching, paged-attention, quantization

**Comment by zhangjun** [comment]
# MeMo
[megatron_lm_encoder_decoder_model.py](https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py)

**Comment by zhangjun** [comment]
Smooth-Quant int8 quantization achieves 2.46x faster latency compared to fp16 mode. (OPT-13B on A100 80GB)

**Comment by zhangjun** [comment]
# Fine Tune

## Prompt Series
[P-Tuning V2](https://zhuanlan.zhihu.com/p/664801668)
[P-Tuning v2: Prompt Tuning Can Be Comparable to Fine-tuning Universally Across Scales and Tasks](https://arxiv.org/pdf/2110.07602.pdf)
[The Power of Scale for Parameter-Efficient Prompt Tuning](https://arxiv.org/pdf/2104.08691.pdf)
[ P-Tuning GPT Understands, Too](https://arxiv.org/pdf/2103.10385.pdf)
## Lora Series
[LORA: LOW-RANK ADAPTATION OF LARGE LANGUAGE MODELS](https://arxiv.org/pdf/2106.09685.pdf)
## SFT
[大模型SFT微调指令数据的生成](https://zhuanlan.zhihu.com/p/650596719)
